src/mapgen.cpp

```diff
@@ -231,6 +231,149 @@ static constexpr int MON_RADIUS = 3;
+    tripoint_bub_ms last_tile = current_tile;
+
+    for( int i = 0; i < streak_length; i++ ) {
+        tripoint_bub_ms destination_tile = get_point_from_direction( streak_direction, last_tile );
+
+        if( !tile_can_have_blood( md, destination_tile, wall_streak ) ) {
+            // We hit a non-valid tile. Try to find a new direction otherwise just terminate the streak.
+            bool terminate_streak = true;
+
+            for( int ii = 1; ii < 5; ii++ ) {
+                if( ii == streak_direction ) {
+                    // We don't want to check the direction we came from. No turning around!
+                    continue;
+                }
+                tripoint_bub_ms adjacent_tile = get_point_from_direction( ii, last_tile );
+
+                if( tile_can_have_blood( md, adjacent_tile, wall_streak ) ) {
+                    streak_direction = ii;
+                    destination_tile = adjacent_tile;
+                    terminate_streak = false;
+                    break;
+                }
+            }
+
+            if( terminate_streak ) {
+                break;
+            }
+        }
+
+        if( rng( 1, 100 ) < 5 + i * 3 ) {
+            // Sometimes a streak should skip a tile, the chance increases with each step
+            last_tile = destination_tile;
+            continue;
+        }
+
+        if( rng( 1, 100 ) < 10 + i * 5 ) {
+            // Sometimes a streak should end early, the chance increases with each step.
+            // This is just a hack to further weight the distribution in favor of short streaks over long ones.
+            break;
+        }
+
+        md.add_field( destination_tile, field_fd_blood );
+        last_tile = destination_tile;
+
+
+        if( ( rng( 1, 100 ) < 30 + i * 3 ) && !wall_streak ) {
+            // Floor streaks can meander and the probability of meandering increases with each step
+            // Long straight streaks aren't visually interesting. So sometimes a streak will curve by meandering to the side.
+
+            int new_direction = rng( 0, 1 );
+            int meander_mod = ( streak_direction >= 3 ) ? 1 : 3;
+            tripoint_bub_ms adjacent_tile = get_point_from_direction( meander_mod + new_direction, last_tile );
+            if( tile_can_have_blood( md, adjacent_tile, false ) ) {
+                md.add_field( adjacent_tile, field_fd_blood );
+                last_tile = adjacent_tile;
+            }
+        }
+    }
+}
+
+static void place_bool_pools( map &md, const tripoint_bub_ms &current_tile )
+{
+    if( !tile_can_have_blood( md, current_tile, false ) ) {
+        // Quick check the first tile is valid for placement
+        return;
+    }
+
+    md.add_field( current_tile, field_fd_blood );
+    place_blood_on_adjacent( md, current_tile, 60 );
+
+    for( int i = 1; i < 4; i++ ) {
+        tripoint_bub_ms adjacent_tile = get_point_from_direction( i, current_tile );
+        place_blood_on_adjacent( md, adjacent_tile, 30 );
```

Comment:
you probably want another `tile_can_have_blood` check here?

Right now the logic is:
- first generation: check and place initial pool
- second generation A (inside `place_blood_on_adjacent`): check and place pools around initial one with some chance
- second generation B (this line): on every tile around the first generation pool: call `place_blood_on_adjacent` (without checks). No pools placed
- third generation (inside `place_blood_on_adjacent`):  on every tile around the second generation B, check and place blood with some chance.

So basically right now the pools can effectively get over walls. Although not very likely and perhaps no big deal